packages/flutter/lib/src/widgets/system_context_menu.dart

```diff
@@ -88,26 +106,67 @@ class SystemContextMenu extends StatefulWidget {
     return MediaQuery.maybeSupportsShowingSystemContextMenu(context) ?? false;
   }
 
+  /// Returns true only if the given list contains at least one pair of
+  /// identical IOSSystemContextMenuItems.
+  static bool _containsDuplicates(List<IOSSystemContextMenuItem> items) {
+    final Set<IOSSystemContextMenuItem> uniqueItems = <IOSSystemContextMenuItem>{};
+
+    for (final IOSSystemContextMenuItem item in items) {
+      if (uniqueItems.contains(item)) {
+        return true;
+      }
+      uniqueItems.add(item);
+    }
+
+    return false;
+  }
+
   @override
   State<SystemContextMenu> createState() => _SystemContextMenuState();
 }
 
 class _SystemContextMenuState extends State<SystemContextMenu> {
   late final SystemContextMenuController _systemContextMenuController;
 
+  /// Return the IOSSystemContextMenuItemData for the given
+  /// IOSSystemContextMenuItem.
+  ///
+  /// IOSSystemContextMenuItem is a format that is designed to be consumed as
+  /// SystemContextMenu.items, where users might want a default localized title
+  /// to be set for them.
+  ///
+  /// IOSSystemContextMenuItemData is a format that is meant to be consumed by
+  /// SystemContextMenuController.show, where there is no expectation that
+  /// localizations can be used under the hood.
+  IOSSystemContextMenuItemData _itemToData(
+    IOSSystemContextMenuItem item,
+    WidgetsLocalizations localizations,
+  ) {
+    return switch (item) {
+      IOSSystemContextMenuItemCut() => const IOSSystemContextMenuItemDataCut(),
+      IOSSystemContextMenuItemCopy() => const IOSSystemContextMenuItemDataCopy(),
+      IOSSystemContextMenuItemPaste() => const IOSSystemContextMenuItemDataPaste(),
+      IOSSystemContextMenuItemSelectAll() => const IOSSystemContextMenuItemDataSelectAll(),
+      IOSSystemContextMenuItemLookUp() => IOSSystemContextMenuItemDataLookUp(
+        title: item.title ?? localizations.lookUpButtonLabel,
+      ),
+      IOSSystemContextMenuItemSearchWeb() => IOSSystemContextMenuItemDataSearchWeb(
+        title: item.title ?? localizations.searchWebButtonLabel,
+      ),
+      IOSSystemContextMenuItemShare() => IOSSystemContextMenuItemDataShare(
+        title: item.title ?? localizations.shareButtonLabel,
+      ),
+      IOSSystemContextMenuItemCustom() => IOSSystemContextMenuItemDataCustom(
+        title: item.title!,
+        onPressed: item.onPressed!,
+      ),
+    };
+  }
+
   @override
   void initState() {
     super.initState();
     _systemContextMenuController = SystemContextMenuController(onSystemHide: widget.onSystemHide);
-    _systemContextMenuController.show(widget.anchor);
```

Comment:
Would it make sense to instantiate the controller with `items`, and then `show` can just take a `targetRect`? Not sure if this is better but wanted to throw it out there.

Reply:
Thanks for bringing this up, I had forgotten about this pattern from ContextMenuController. I've investigated and I think I should keep `items` in `show`, though. The way it's used in _SystemContextMenuControllerState works nicely with the items passed in at show time. It seems like I ended up changing ContextMenuController to this pattern as well during code review ([code](https://github.com/flutter/flutter/blob/06c3c93d72e1a632cf5dd61664a41b32d950c3de/packages/flutter/lib/src/widgets/context_menu_controller.dart#L32)).